### PR TITLE
soc: arm: atmel: Add SAM3U series

### PR DESCRIPTION
**Pull request by xingrz**
This commit preliminarily adds SoC support for SAM3U series.

It also introduces a development board ATSAM3U-EK.

Datasheet: https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-6430-32-bit-Cortex-M3-Microcontroller-SAM3U4-SAM3U2-SAM3U1_Datasheet.pdf

### CODE REVIEW

**Comment by nandojve** [comment]
Hi @xingrz ,

To add a series it is necessary add a board which is available, in general a devkit. Try focus in bring board up with serial console. Remaining peripheral should be enabled one by one to make sure we keep compatibility and avoid bring too many maintainers in the same PR. Remember, small changes fast merge.

> USB device driver is still missing for the entire Atmel SAM family. I'll try to do it in another PR.

Currently there are two USB drives for SAM family but those are not compatible with UDP. If you can work on it will be awesome! The USB should be in another PR for sure.

Remember that bring a new series or complex drivers like USB takes time to review, be patient.

**Comment by xingrz** [comment]
Hi @nandojve , Thanks for your advice. In fact, I didn't own any official devkit boards yet. I'm doing my development base on the reverse engineering of an existing product (a very common SAM3U2C based DAPLink debugger). Is this board acceptable?

Currently, I've been able to set up a working UART console and get the LEDs to blink. Since the SAM3U shares many IPs with the existing SAM3X (with only the regulator addresses being different), I believe that it won't take too much effort to get everything working.

**Comment by nandojve** [comment]
Hi @xingrz ,

The development reference for this SoC is [ATSAM3U-EK](https://www.microchip.com/en-us/development-tool/ATSAM3U-EK) board. You should introduce this board if you want have SAM3U on Zephyr.

> Since the SAM3U shares many IPs with the existing SAM3X (with only the regulator addresses being different), I believe that it won't take too much effort to get everything working.

This is not 100% true. Yes, those series shares a lot but sometimes the IP is not 100% equal. The USB for instance, in the SAM3U is a UDPHS which only exists on this Series. The SAM3X uses UOTGHS, SAM4 UDP, SAM4L USBC and SAMV7 USBHS, SAMG UDP+UHP so there are at least 6 different implementations in the whole series.

Since drivers uses registers definitions and it is quite normal to find problems with descriptions and missing entries it realy necessary to have a board in the tree. This way it is possible to keep the ASF library update and IP share is effective.

**Comment by xingrz** [comment]
Got it! I'll update this PR later. Thanks for your advice!

**Comment by xingrz** [comment]
Hi @nandojve, I've added the ATSAM3U-EK board and confirmed that both samples/basic/blinky and samples/basic/button are working, including the serial logging.

**Comment by nandojve** [comment]
Hi @xingrz,

As you can check at _Manifest / Unit Test Results (pull_request_target)_ some tests related to flash driver are failing. What you need to do is check if EEFC peripheral is compatible with current driver. If yes, the registers macros should be adjusted at **hal_atmel**. Otherwise, we need to make flash driver compatible.

As you can see, build a board is very important to get those details.

Note, this PR could be delayed by #55655, it is not clear yet.

**Comment by xingrz** [comment]
> As you can check at _Manifest / Unit Test Results (pull_request_target)_ some tests related to flash driver are failing. What you need to do is check if EEFC peripheral is compatible with current driver. If yes, the registers macros should be adjusted at **hal_atmel**. Otherwise, we need to make flash driver compatible.

Thank you for letting me know. I've looked into the EEFC peripheral and found that it's similar to the one on SAM4E, but the `EEFC_FSR` on SAM3U doesn't have the `FLERR` bit, which caused the compilation error earlier. Additionally, since SAM3U lacks the command to erase flash by page, I will be disabling the flash driver for SAM3U, following the same approach that had been used for SAM3X. This should prevent the compilation error from occurring again.

> Note, this PR could be delayed by #55655, it is not clear yet.

Okay, I'll follow up with you.

**Comment by nandojve** [comment]
Hi @xingrz ,

Thank you for investigate the issue.

> Additionally, since SAM3U lacks the command to erase flash by page, I will be disabling the flash driver for SAM3U, following the same approach that had been used for SAM3X. This should prevent the compilation error from occurring again.

This approach is OK. You don't need provide a flash driver to introduce this series.

File: boards/arm/sam3u_ek/Kconfig.board
Comment (nandojve) [inline boards/arm/sam3u_ek/Kconfig.board:8]: This commit should be split in:

1) only west.yml (it should be first in the whole series)
2) dts folder entries
3) soc folder content
4) drivers/<any>
5) board

File: soc/arm/atmel_sam/common/soc_pmc.c
Comment (nandojve) [inline soc/arm/atmel_sam/common/soc_pmc.c:3]: This commit should be introduced after soc commit and before drivers.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @xingrz ,

A few organizational changes.

File: dts/arm/atmel/sam3u.dtsi
Comment (nandojve) [inline dts/arm/atmel/sam3u.dtsi:34]: Hi @xingrz ,

Please, rebase your changes on top of main and update your PR to use clocks instead peripheral-id.

**Review by nandojve — CHANGES_REQUESTED** [review]

File: boards/arm/sam3u_ek/doc/index.rst
Comment (nandojve) [inline boards/arm/sam3u_ek/doc/index.rst:38]: Let's keep SAM peripheral on alphabetic order.

File: boards/arm/sam3u_ek/sam3u_ek_defconfig
Comment (nandojve) [inline boards/arm/sam3u_ek/sam3u_ek_defconfig:10]:
```diff
CONFIG_SOC_SERIES_SAM3U=y
CONFIG_SOC_PART_NUMBER_SAM3U4C=y
+CONFIG_BOARD_SAM3U_EK=y
CONFIG_ARM_MPU=y
CONFIG_CONSOLE=y
CONFIG_UART_CONSOLE=y
CONFIG_SERIAL=y
-CONFIG_BOARD_SAM3U_EK=y
CONFIG_GPIO=y
```

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @xingrz ,

Small nits. After this, we need wait hal be merged to change west.yml file. After that, PR is ready.

File: drivers/hwinfo/hwinfo_sam.c
Comment (pdgendt) [inline drivers/hwinfo/hwinfo_sam.c:78]: Analogues to the discussion in #53200
```suggestion
#if !defined(CONFIG_SOC_SERIES_SAM3U) && !defined(CONFIG_SOC_SERIES_SAM3X)
```

File: dts/arm/atmel/sam3u.dtsi
Comment (pdgendt) [inline dts/arm/atmel/sam3u.dtsi:82]: The `hwinfo` driver doesn't use this device tree node, if it can't be enabled should we move it outside of this file?

File: drivers/hwinfo/hwinfo_sam.c
Comment (xingrz) [reply drivers/hwinfo/hwinfo_sam.c]: Thanks. Updated.

File: dts/arm/atmel/sam3u.dtsi
Comment (xingrz) [reply dts/arm/atmel/sam3u.dtsi:82]: Actually it [does use it](https://github.com/zephyrproject-rtos/zephyr/blob/f423ec255fb76641f69a65d1e57f3c6dde70b15d/drivers/hwinfo/hwinfo_sam.c#L36). I followed the same approach as [sam3x did](https://github.com/zephyrproject-rtos/zephyr/blob/f423ec255fb76641f69a65d1e57f3c6dde70b15d/dts/arm/atmel/sam3x.dtsi#L63).

**Comment by xingrz** [comment]
Hi @nandojve,

Given that #57400 was rejected, is it still possible to move forward with this PR?

**Comment by nandojve** [comment]
Hi @xingrz ,

Sincerely, I don't know and I think when _it is interesting for the project_ they create an exception. Right now the only way, as far I understood, is submit the **bsd-atmel** to the opensource [review-process](https://opensource.org/licenses/review-process). However, if it is approved, for sure, there is no more limitation.

No guarantee of success but nothing to lose : )

**Comment by xingrz** [comment]
@nandojve Anyway, this is my first time submitting SoC support, and I have learned a lot from the experience. Thank you!

I will close this PR for now. I'll make it as an out-of-tree support in the future.

**Pull request closed without merging**
